afkak/brokerclient.py

```diff
@@ -319,13 +319,15 @@ def _sendQueued(self):
     # FIXME: This should not be a public API. The Deferred returned by
     # makeRequest already has a cancel() method.
     # XXX: Wat is the fourth argument I don't even?!
-    def cancelRequest(self, requestId, reason=CancelledError(), _=None):
```

Comment:
The 4th arg is there because of the way this is called via a `Partial()` wrapper and because `deferred.cancel()` passes 'self' to the canceller function which we don't care about, we only care about the stuff the partial wrapper has stuffed into it.

Reply:
There could certainly be a better way though :-)

Reply:
I was mostly reacting to the *name* of the parameter... `_` is not a great name, particularly for a public method.

I will say that how cancellation works in this class isn't quite right because it is used both by the consumer of the deferred and the producer (`BrokerClient`). Deferred really has _two_ APIs:

- Producer: `__init__()`, `callback()`, `errback()`. The code that instantiates the deferred is responsible for populating it with a value.
- Consumer: `addCallbacks()` et. al., **`cancel()`**. For the code that uses the value the deferred represents.

`BrokerClient` calls `cancel()` internally in some circumstances, violating this separation. That means that if the consumer actually tries to call `cancel()` it will throw!